### PR TITLE
fix(eviction): Tune eviction threshold in cache mode

### PR DESCRIPTION
**Pull request by BagritsevichStepan**
fixes dragonflydb#4139

### CODE REVIEW

File: tests/dragonfly/generic_test.py [hunk 1]
```diff
@@ -149,7 +149,7 @@ async def test_reply_guard_oom(df_factory, df_seeder_factory):
 @pytest.mark.asyncio
 async def test_denyoom_commands(df_factory):
     df_server = df_factory.create(
-        proactor_threads=1, maxmemory="256mb", oom_deny_commands="get", oom_deny_ratio=0.7
+        proactor_threads=1, maxmemory="256mb", oom_deny_commands="get", rss_oom_deny_ratio=0.7
```
Comment (adiholden) [inline tests/dragonfly/generic_test.py:152]: you should not replace oom_deny_ratio with  rss_oom_deny_ratio in this test
This test checks that we return oom for denyoom commands
so now we dont have the ration anymore we will return oom when we reach maxmemory set by user

Comment (BagritsevichStepan) [reply tests/dragonfly/generic_test.py:152]: Yes, will fix

File: src/server/dragonfly_test.cc [hunk 2]
```diff
@@ -26,7 +26,7 @@ ABSL_DECLARE_FLAG(float, mem_defrag_threshold);
 ABSL_DECLARE_FLAG(float, mem_defrag_waste_threshold);
 ABSL_DECLARE_FLAG(uint32_t, mem_defrag_check_sec_interval);
 ABSL_DECLARE_FLAG(std::vector<std::string>, rename_command);
-ABSL_DECLARE_FLAG(double, oom_deny_ratio);
+ABSL_DECLARE_FLAG(double, rss_oom_deny_ratio);
```
Comment (adiholden) [inline src/server/dragonfly_test.cc:29]: nit:
i dont see you use this flag in this file

File: src/server/engine_shard.cc [hunk 3]
```diff
@@ -717,8 +723,17 @@ void EngineShard::RetireExpiredAndEvict() {
     ttl_delete_target = kTtlDeleteLimit * double(deleted) / (double(traversed) + 10);
   }
 
-  ssize_t eviction_redline = size_t(max_memory_limit * kRedLimitFactor) / shard_set->size();
+  const size_t max_rss_memory =
+      max_memory_limit * std::max(ServerState::tlocal()->rss_oom_deny_ratio, 0.0);
+  const ssize_t rss_memory =
+      ConvertNumericOrRound<uint64_t, ssize_t>(rss_mem_current.load(memory_order_relaxed));
```
Comment (adiholden) [inline src/server/engine_shard.cc:729]: why do you need this ConvertNumericOrRound?

File: src/server/engine_shard.cc [hunk 4]
```diff
@@ -704,6 +709,7 @@ void EngineShard::RetireExpiredAndEvict() {
   { std::unique_lock lk(db_slice.GetSerializationMutex()); }
   constexpr double kTtlDeleteLimit = 200;
   constexpr double kRedLimitFactor = 0.1;
```
Comment (adiholden) [inline src/server/engine_shard.cc:711]: rename to kEvictionFreeMemoryThreshold
add a comment when shard free memory goes below max_memory*threshold we start evicting

File: src/server/engine_shard.cc [hunk 5]
```diff
@@ -717,8 +723,17 @@ void EngineShard::RetireExpiredAndEvict() {
     ttl_delete_target = kTtlDeleteLimit * double(deleted) / (double(traversed) + 10);
   }
 
-  ssize_t eviction_redline = size_t(max_memory_limit * kRedLimitFactor) / shard_set->size();
+  const size_t max_rss_memory =
+      max_memory_limit * std::max(ServerState::tlocal()->rss_oom_deny_ratio, 0.0);
```
Comment (adiholden) [inline src/server/engine_shard.cc:727]: if rss_oom_deny_ratio = -1 than max_rss_memory is 0?
I believe we should write
rss_limit = ServerState::tlocal()->rss_oom_deny_ratio > 0? max_memory_limit * ServerState::tlocal()->rss_oom_deny_ratio :  std::numeric_limits<size_t>::max()

File: src/server/engine_shard.cc [hunk 6]
```diff
@@ -736,10 +751,12 @@ void EngineShard::RetireExpiredAndEvict() {
     }
 
     // if our budget is below the limit
-    if (db_slice.memory_budget() < eviction_redline && GetFlag(FLAGS_enable_heartbeat_eviction)) {
+    if ((db_slice.memory_budget() < memory_budget_threshold || rss_memory > rss_memory_threshold) &&
+        GetFlag(FLAGS_enable_heartbeat_eviction)) {
       uint32_t starting_segment_id = rand() % pt->GetSegmentCount();
-      db_slice.FreeMemWithEvictionStep(i, starting_segment_id,
-                                       eviction_redline - db_slice.memory_budget());
+      size_t goal_bytes = std::max(memory_budget_threshold - db_slice.memory_budget(),
```
[the remark this answers is not in the document]
Comment (BagritsevichStepan) [reply src/server/engine_shard.cc:757]: ssize_t*

**Comment by BagritsevichStepan** [comment]
@adiholden I believe I've identified the issue: I'm calculating rss_threshold for each shard individually, but the current rss_memory value is a global metric for all shards

File: src/server/engine_shard.cc [hunk 7]
```diff
@@ -716,7 +742,12 @@ void EngineShard::RetireExpiredAndEvict() {
     ttl_delete_target = kTtlDeleteLimit * double(deleted) / (double(traversed) + 10);
   }
 
-  ssize_t eviction_redline = size_t(max_memory_limit * kRedLimitFactor) / shard_set->size();
+  // For memory we are using threshold for free memory
+  const ssize_t shard_memory_budget_threshold =
+      ssize_t(max_memory_limit * kEvictionFreeMemoryThreshold) / shard_set->size();
+
+  // Calculate how much rss memory we need to free per shard
+  ssize_t rss_memory_to_free = CalculateRssMemoryToFreePerShard();
```
Comment (adiholden) [inline src/server/engine_shard.cc:750]: in the for loop below we call DeleteExpiredStep which can also remove entries, so calculating the rss memory to free here and updating it with released_bytes is not enough.

File: src/server/engine_shard.cc [hunk 8]
```diff
@@ -735,10 +766,17 @@ void EngineShard::RetireExpiredAndEvict() {
...
 
     // if our budget is below the limit
-    if (db_slice.memory_budget() < eviction_redline && GetFlag(FLAGS_enable_heartbeat_eviction)) {
+    if ((db_slice.memory_budget() < shard_memory_budget_threshold || rss_memory_to_free > 0) &&
+        GetFlag(FLAGS_enable_heartbeat_eviction)) {
       uint32_t starting_segment_id = rand() % pt->GetSegmentCount();
-      db_slice.FreeMemWithEvictionStep(i, starting_segment_id,
-                                       eviction_redline - db_slice.memory_budget());
+      size_t goal_bytes =
+          std::max(rss_memory_to_free, shard_memory_budget_threshold - db_slice.memory_budget());
```
Comment (adiholden) [inline src/server/engine_shard.cc:773]: I see now that also the old logic for multiple dbs is broken because memory budget is update only when CacheStats is called (once in heartbeat)

File: src/server/engine_shard.cc [hunk 9]
```diff
@@ -198,6 +198,33 @@ optional<uint32_t> GetPeriodicCycleMs() {
   return clock_cycle_ms;
 }
 
+/* Eviction begins when the shard's free memory falls below (max_memory_limit *
+ * kEvictionFreeMemoryThreshold) / <number of shards> */
+constexpr double kEvictionFreeMemoryThreshold = 0.1;
+/* or eviction starts if the used RSS memory exceeds max_memory_limit * kEvictionRssThreshold */
+constexpr double kEvictionRssThreshold = 1.0 - kEvictionFreeMemoryThreshold;
```
Comment (adiholden) [inline src/server/engine_shard.cc:205]: I think this constant is kind of redundant
you can do the same for rss as we do for used memory
rss_limit = rss_oom_deny_ratio * max_memory_limit
eviction_rss_threshold = rss_limit * kEvictionFreeMemoryThreshold
curr_free_rss = rss_limit - rss_mem_current
if (curr_free_rss < eviction_rss_threshold) {
 return eviction_rss_threshold - curr_free_rss
}
return 0

File: src/server/engine_shard.cc (hunk 7 above)
Comment (BagritsevichStepan) [reply src/server/engine_shard.cc:750]: Yes, I understand it

File: tests/dragonfly/memory_test.py [hunk 10]
```diff
@@ -107,3 +107,74 @@ async def test_rss_oom_ratio(df_factory: DflyInstanceFactory, admin_port):
...
+    key_size = 1024  # 1 mb
+    num_keys = data_fill_size // key_size
+
+    # Fill data to 25% of max memory
+    await async_client.execute_command("DEBUG", "POPULATE", num_keys, "key", key_size)
+
+    await asyncio.sleep(1)  # Wait for RSS heartbeat update
+
+    # First test that eviction is not triggered without connection creation
+    stats_info = await async_client.info("stats")
```
Comment (adiholden) [inline tests/dragonfly/memory_test.py:142]: lets also add a check here that used memory is below maxmemory * 0.9 and rss curr memory is below 
maxmemory * rss_oom_deny_ratio * 0.9

File: tests/dragonfly/memory_test.py [hunk 11]
```diff
@@ -107,3 +107,74 @@ async def test_rss_oom_ratio(df_factory: DflyInstanceFactory, admin_port):
...
+    assert stats_info["evicted_keys"] == 0, "No eviction should start yet."
+
+    # Get RSS memory before creating new connections
+    memory_info = await async_client.info("memory")
+    rss_before_connections = memory_info["used_memory_rss"]
+
+    # Increase RSS memory by 30% of max memory
+    # We can simulate RSS increase by creating new connections
+    # Estimate memory per connection
+    estimated_connection_memory = 15 * 1024  # 15 KB per connection
```
Comment (adiholden) [inline tests/dragonfly/memory_test.py:152]: To make this test supper stable and more controlled
I would run
CONFIG SET enable_heartbeat_eviction false
before creating the connections
after that check the rss make sure its above the threshold
then run 
CONFIG SET enable_heartbeat_eviction true
and after that check that rss memory dropped below the threshold and check the stats for evicted keys as you did below

File: src/server/dragonfly_test.cc [hunk 12]
```diff
@@ -456,19 +455,19 @@ TEST_F(DflyEngineTest, OOM) {
...
   shard_set->TEST_EnableCacheMode();
 
   ssize_t i = 0;
   RespExpr resp;
+  std::string value(1024, 'b');
   for (; i < 10000; ++i) {
-    resp = Run({"setex", StrCat("key", i), "30", "bar"});
+    resp = Run({"setex", StrCat("key", i), "30", value});
     // we evict some items because 5000 is too much when max_memory_limit is 300000.
+    if (resp != "OK") {
```
[the remark this answers is not in the document]
Comment (BagritsevichStepan) [reply src/server/dragonfly_test.cc]: No more needed. Just increased memory here. Because sometimes it fails. Its can be faster than heartbeat

File: src/server/engine_shard.cc [hunk 13]
```diff
@@ -198,6 +202,30 @@ optional<uint32_t> GetPeriodicCycleMs() {
   return clock_cycle_ms;
 }
 
+ssize_t CalculateMemoryBudget(size_t max_memory, size_t used_memory) {
+  if (max_memory >= used_memory) {
+    return max_memory - used_memory;
+  }
+  // negative value indicates memory overuse
+  return -1 * ssize_t(used_memory - max_memory);
```
Comment (BorysTheDev) [inline src/server/engine_shard.cc:210]: UB because size_t - size_t = size_t

Comment (BagritsevichStepan) [reply src/server/engine_shard.cc:210]:
```
ssize_t CalculateMemoryBudget(size_t max_memory, size_t used_memory) {
  if (max_memory >= used_memory) {
    return max_memory - used_memory;
  }
  // negative value indicates memory overuse
  return -1 * ssize_t(used_memory - max_memory);
}
```

Comment (BorysTheDev) [reply src/server/engine_shard.cc:210]: Ok I got, the result is correct because you use overflow of ssize_t type, but it can be UB I'm not sure. please rewrite

File: src/server/engine_shard.cc [hunk 14]
```diff
@@ -198,6 +202,30 @@ optional<uint32_t> GetPeriodicCycleMs() {
...
+}
+
+ssize_t CalculateFreeMemoryOnShard() {
+  // Calculate how much memory is used by all shards
+  const size_t current_global_rss_memory = used_mem_current.load(memory_order_relaxed);
+  return CalculateMemoryBudget(max_memory_limit, current_global_rss_memory) / shard_set->size();
```
Comment (BorysTheDev) [inline src/server/engine_shard.cc:216]: you again use ssize_t overflow that can be UB

File: src/server/engine_shard.cc (hunk 13 above)
Comment (BagritsevichStepan) [reply src/server/engine_shard.cc:210]: Do you mean to round size_t to max ssize_t? Because I already implemented this previously
https://github.com/dragonflydb/dragonfly/pull/4142#discussion_r1858484944

Comment (BorysTheDev) [reply src/server/engine_shard.cc:210]: just update ssize_t CalculateMemoryBudget(ssize_t max_memory, ssize_t used_memory)

Comment (BagritsevichStepan) [reply src/server/engine_shard.cc:210]: But we are still converting size_t to ssize_t before calling the function. This is even more dangerous because we did not perform the subtraction.

Comment (BorysTheDev) [reply src/server/engine_shard.cc:210]: in our case, such converting is absolutely safe, because we will never get such a memory size  in the near future

**Review by BorysTheDev — DISMISSED** [review]

**Comment by BagritsevichStepan** [comment]
Upd.:
1. Fixed a bug where using ssize_t still caused overflow. I refactored the code and completely removed the use of ssize_t in this context.
2. Identified why `DflyEngineTest.Bug207` was failing. The issue was that we were only evicting items during `AddOrFindInternal`. During each heartbeat, we expected to evict a significant amount of data but successfully evicted 0 bytes. This happened because the strings were small, inlined, and HasAllocated returned false for them (this issue also exists in the main branch). To fix this, I increased the length of the values in the test.
3. Added additional logs for better debugging and monitoring.

**Review by BorysTheDev — DISMISSED** [review]

File: src/server/db_slice.cc [hunk 15]
```diff
@@ -1291,9 +1291,6 @@ pair<uint64_t, size_t> DbSlice::FreeMemWithEvictionStep(DbIndex db_ind, size_t s
 
     auto time_finish = absl::GetCurrentTimeNanos();
     events_.evicted_keys += evicted_items;
-    DVLOG(2) << "Evicted: " << evicted_bytes;
```
Comment (adiholden) [inline src/server/db_slice.cc:1294]: why did you remove this print?

Comment (BagritsevichStepan) [reply src/server/db_slice.cc:1294]: Moved this to the [Heartbeat](https://github.com/dragonflydb/dragonfly/pull/4142/files#diff-89b55bcf480e65db9741131907708968288c37fd0993ac1680ba2eb7fb28e873R804)

Comment (BagritsevichStepan) [reply src/server/db_slice.cc:1294]: Because this callback is not always called

File: src/server/engine_shard.cc [hunk 16]
```diff
@@ -716,7 +732,33 @@ void EngineShard::RetireExpiredAndEvict() {
     ttl_delete_target = kTtlDeleteLimit * double(deleted) / (double(traversed) + 10);
   }
 
-  ssize_t eviction_redline = size_t(max_memory_limit * kRedLimitFactor) / shard_set->size();
+  const size_t shards_count = shard_set->size();
+  const double eviction_memory_budget_threshold = GetFlag(FLAGS_eviction_memory_budget_threshold);
+  const double rss_oom_deny_ratio = ServerState::tlocal()->rss_oom_deny_ratio;
+
+  const bool should_evict =
+      eviction_memory_budget_threshold > 0.0 && GetFlag(FLAGS_enable_heartbeat_eviction);
```
Comment (adiholden) [inline src/server/engine_shard.cc:740]: why do you check eviction_memory_budget_threshold > 0.0 ?
nit: maybe rename this var should_evict -> eviction_enabled

File: src/server/engine_shard.cc [hunk 17]
```diff
@@ -734,11 +776,35 @@ void EngineShard::RetireExpiredAndEvict() {
...
+
+    const size_t global_used_memory = used_mem_current.load(memory_order_relaxed);
+
+    // Calculate how many bytes we need to evict on this shard
+    size_t goal_bytes = CalculateHowManyBytesToEvictOnshard(max_memory_limit, global_used_memory,
+                                                            shard_memory_budget_threshold);
+
+    if (should_evict_depending_on_rss_memory) {
+      // Calculate how much rss memory is used by all shards
+      io::Result<io::StatusData> sdata_res = io::ReadStatusInfo();
```
Comment (adiholden) [inline src/server/engine_shard.cc:790]: this RetireExpiredAndEvict function runs in the heart beat flow that runs every 10ms
today UpdateMemoryGlobalStats runs every 100ms, and I believe there is a reason for this, not to read the rss data so many time. In this logic now you run this read several times in this heartbeat flow (as the count of dbs) and I dont think we want this.
Further more you use above the used_mem_current that is updated in the UpdateMemoryGlobalStats and will not get updated in this for loop

Comment (BagritsevichStepan) [reply src/server/engine_shard.cc:790]: But used_mem_current can be updated by other threads

File: src/server/db_slice.cc (hunk 15 above)
Comment (adiholden) [reply src/server/db_slice.cc:1294]: oh I see you moved the prints to engine shard. ok :+1:

File: src/server/engine_shard.cc (hunk 17 above)
Comment (adiholden) [reply src/server/engine_shard.cc:790]: I understand. The flow is indead very confusing!
I suggest to make this simpler by calculating the bytes goal to evict before the forloop on the dbs and once you enter the for loop you have the goal and you update it with the returned evicted byts value from FreeMemWithEvictionStep
In a later stage (not this PR) we can improve to update the rss fetch in this flow so it will be updated in case of eviction step if needed

File: src/server/engine_shard.cc [hunk 18]
```diff
@@ -716,7 +732,33 @@ void EngineShard::RetireExpiredAndEvict() {
...
+  const size_t shards_count = shard_set->size();
+  const double eviction_memory_budget_threshold = GetFlag(FLAGS_eviction_memory_budget_threshold);
+  const double rss_oom_deny_ratio = ServerState::tlocal()->rss_oom_deny_ratio;
+
+  const bool should_evict =
+      eviction_memory_budget_threshold > 0.0 && GetFlag(FLAGS_enable_heartbeat_eviction);
+  // If rss_oom_deny_ratio is set, we should evict depending on rss memory too
+  const bool should_evict_depending_on_rss_memory = should_evict && rss_oom_deny_ratio > 0.0;
+
+  size_t shard_memory_budget_threshold;
```
Comment (adiholden) [inline src/server/engine_shard.cc:744]: In addition I believe that it will be more readable if you write a function that will return the byte goal for eviction in this heartbeat invocation. It will have inside all this new logic that you introduced and call this function only if eviction is enabled

File: src/server/engine_shard.cc (hunk 16 above)
Comment (BagritsevichStepan) [reply src/server/engine_shard.cc:740]: What does it mean when `eviction_memory_budget_threshold` is less than 0? I believe it indicates that this flag is disabled

File: src/server/db_slice.cc [hunk 19]
```diff
@@ -1217,6 +1217,7 @@ auto DbSlice::DeleteExpiredStep(const Context& cntx, unsigned count) -> DeleteEx
       auto prime_it = db.prime.Find(it->first);
       CHECK(!prime_it.is_done());
       ExpireIfNeeded(cntx, prime_it, false);
+      result.deleted_bytes += prime_it->second.MallocUsed();
```
[the remark this answers is not in the document]
Comment (BagritsevichStepan) [reply src/server/db_slice.cc:1219]: yes, already fixed

[the remark this answers is not in the document]
Comment (adiholden) [reply src/server/db_slice.cc:1219]: also you can add the key malloc used to deleted bytes

File: src/server/db_slice.cc [hunk 20]
```diff
@@ -1216,6 +1216,7 @@ auto DbSlice::DeleteExpiredStep(const Context& cntx, unsigned count) -> DeleteEx
     if (ttl <= 0) {
       auto prime_it = db.prime.Find(it->first);
       CHECK(!prime_it.is_done());
+      result.deleted_bytes += prime_it->second.MallocUsed();
```
Comment (adiholden) [inline src/server/db_slice.cc:1219]: also you can add the key malloc used to deleted bytes

File: src/server/engine_shard.cc [hunk 21]
```diff
@@ -716,11 +771,11 @@ void EngineShard::RetireExpiredAndEvict() {
     ttl_delete_target = kTtlDeleteLimit * double(deleted) / (double(traversed) + 10);
   }
 
-  ssize_t eviction_redline = size_t(max_memory_limit * kRedLimitFactor) / shard_set->size();
-
   DbContext db_cntx;
   db_cntx.time_now_ms = GetCurrentTimeMs();
 
+  size_t goal_bytes = GetGoalBytesToEvict();
```
Comment (adiholden) [inline src/server/engine_shard.cc:777]: nit:
goal_bytes -> eviction_bytes / eviction_goal
GetGoalBytesToEvict -> CalculateEvictionByets

**Comment by BagritsevichStepan** [comment]
:face_with_head_bandage:

File: src/server/engine_shard.cc (hunk 16 above)
Comment (adiholden) [reply src/server/engine_shard.cc]: I dont think we need this logic to disable the flag as we already have enable_heartbeat_eviction flag that disables/enables the heartbeat , and this logic with eviction_memory_budget_threshold > 0 is redundant.
In general if you create such logic to a flag than document it in the flag string description

**Comment by BagritsevichStepan** [comment]
Can not reproduce it

**Review by adiholden — APPROVED** [review]
